### DISCUSSION TITLE
Can Tycho build for a different os, windowing system, architecture?

### DISCUSSION THREAD

**Discussion by jclausius-sg** [eclipse-tycho/tycho#4448] [Q&A] [answered]
Is Tycho (4.0.9) capable of building an RCP (p2 plugin based .product file) for a different system?

I am running maven on a Linux (GTK) system using Tycho 4.0.9.  My base pom.xml has the following:

&lt;plugin&gt;
  &lt;groupId&gt;org.eclipse.tycho&lt;/groupId&gt;
  &lt;artifactId&gt;target-platform-configuration&lt;/artifactId&gt;
  &lt;version&gt;${tycho-version}&lt;/version&gt;
  &lt;configuration&gt;
    &lt;environments&gt;
      &lt;environment&gt;
        &lt;os&gt;win32&lt;/os&gt;
        &lt;ws&gt;win32&lt;/ws&gt;
        &lt;arch&gt;x86_64&lt;/arch&gt;
      &lt;/environment&gt;
    &lt;/environments&gt;
  &lt;/configuration&gt;
&lt;/plugin&gt;

Within my app.product, I have all the various win32 (for 64-bit) plugins listed.  For example:
&lt;plugin id="org.eclipse.equinox.security.win32.x86_64"/&gt;
&lt;plugin id="org.eclipse.core.filesystem.win32.x86_64"/&gt;
etc.

--- when I run maven to build (mvn clean install ...), it generates the following:

[INFO] Resolving dependencies of MavenProject: app.build.product:app.build.product:11.0.0-SNAPSHOT @ /home/build/dev/app/src/app.product/pom.xml
[INFO] {osgi.os=win32, osgi.ws=win32, org.eclipse.update.install.features=true, osgi.arch=x86_64, org.eclipse.update.install.sources=true}
[ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: app.product 11.0.0.qualifier
[ERROR]   Missing requirement: app.product 11.0.0.qualifier requires 'org.eclipse.equinox.p2.iu; org.eclipse.core.filesystem.win32.x86_64 0.0.0' but it could not be found
[ERROR] 
...
[ERROR] Cannot resolve dependencies of project app.build.product:app.build.product:eclipse-repository:11.0.0-SNAPSHOT
[ERROR]  with context {osgi.os=win32, osgi.ws=win32, org.eclipse.update.install.features=true, osgi.arch=x86_64, org.eclipse.update.install.sources=true}
[ERROR]   Software being installed: app.guiclient.product 11.0.0.qualifier
[ERROR]   Missing requirement: app.guiclient.product 11.0.0.qualifier requires 'org.eclipse.equinox.p2.iu; org.eclipse.core.filesystem.win32.x86_64 0.0.0' but it could not be found: See log for details

Is it possible to build an RCP for a different os, windowing system, and architecture (win32, win32, x86_64) from where maven/tycho is run (linux, gtk, x86_64)?

Thanks
-Jeff

**Comment by laeubi** [comment]
You need to list all environments that you like to build in the target configuration, you can see an example here:

https://github.com/eclipse-m2e/m2e-core/blob/3af1fc53e8709cc68d01322d2b06f2d8999bf7f5/pom.xml#L127-L147

then Tycho will pick up what is needed, to ease your setup you should use features as the usually already have these platform dependent things included, e.g. the platform.sdk feature. Then you only need to list your own bundles, all dependencies will be picked up automatically!

**Reply by jclausius-sg** [reply]
I'm a bit confused.

To re-iterate, my pom.xml for target-platform-configuration :
...
&lt;configuration&gt;
  &lt;environments&gt;
    &lt;environment&gt;
      &lt;os&gt;win32&lt;/os&gt;
      &lt;ws&gt;win32&lt;/ws&gt;
      &lt;arch&gt;x86_64&lt;/arch&gt;
    &lt;/environment&gt;
  &lt;/environments&gt;
&lt;/configuration&gt;
...

is configured to target Win32/Win32/x86_64 (as I'd like to generated a Windows binary), but the actual maven/tycho process to build this is being run on a Linux system.

Are you saying, it is something different just because maven/tycho is running on Linux?

-Jeff

**Reply by laeubi** [reply]
You can build the Windows version of your product under Linux. If you running tests it might become a problem but the building itself should work.

**Reply by jclausius-sg** [reply]
There are no test.  Just the build.  Maven with Tycho spits out the following:

[INFO] Resolving dependencies of MavenProject: app.build.product:app.build.product:11.0.0-SNAPSHOT @ /home/build/dev/app/src/app.product/pom.xml
[INFO] {osgi.os=win32, osgi.ws=win32, org.eclipse.update.install.features=true, osgi.arch=x86_64, org.eclipse.update.install.sources=true}
[ERROR] Cannot resolve project dependencies:
[ERROR] Software being installed: app.product 11.0.0.qualifier
[ERROR] Missing requirement: app.product 11.0.0.qualifier requires 'org.eclipse.equinox.p2.iu; org.eclipse.core.filesystem.win32.x86_64 0.0.0' but it could not be found
[ERROR]
...
[ERROR] Cannot resolve dependencies of project app.build.product:app.build.product:eclipse-repository:11.0.0-SNAPSHOT
[ERROR] with context {osgi.os=win32, osgi.ws=win32, org.eclipse.update.install.features=true, osgi.arch=x86_64, org.eclipse.update.install.sources=true}
[ERROR] Software being installed: app.guiclient.product 11.0.0.qualifier
[ERROR] Missing requirement: app.guiclient.product 11.0.0.qualifier requires 'org.eclipse.equinox.p2.iu; org.eclipse.core.filesystem.win32.x86_64 0.0.0' but it could not be found: See log for details

--------------
As mentioned, the environment elements for win32, win32 and x86_64 are properly defined in pom.xml under target-platform-configuration.

It seems tycho is using my OS/environment for trying to resolve a dependency.  Not certain how to override.

**Reply by jclausius-sg** [reply]
For what it's worth, if I download the EXACT same code from our source control and run it in a sample win32, win32, x96_64 environment (using the same versions of OpenJDK and Maven) the maven "clean install" command works without a problem.

Again, my guess is Tycho is looking at the build environment to determine what to download, and not what's specified in the "main" pom.xml.

---
Note, there are no .target files defining anything for a PDE build, would it have anything to do with that?  

Is it possible that even though "target-platform-configuration" is defined in a root level's (or main) pom.xml and there are &lt;modules&gt; used to build sub-directories/sub-projects where the PDE *.product files defined used with their own pom.xml, which also do NOT explicitly define target-platform-configuration the cause of the environment not being found?

-Jeff

**Reply by sratz** [reply, marked as answer]
What Eclipse version are you targeting?
There is no `org.eclipse.core.filesystem.win32.x86_64` bundle anymore since Eclipse 2024-09: https://github.com/eclipse-platform/eclipse.platform/pull/1476

**Reply by jclausius-sg** [reply]
Well...  maybe that's the problem?? 

Eclipse 2024-09 was the IDE that added it there when I clicked the "Add Required plug-ins" from the Product Editor.  However, I did that on Linux.  Let me try the same thing within a Windows version of Eclipse to see if it generates the same list of plug-ins for the .product file.

It'll be coming up soon.
-Jeff

**Reply by jclausius-sg** [reply]
Based on Eclipse in Windows, there were some changes I missed:

Removed the following plug-ins:
&lt;plugin id="org.eclipse.core.filesystem.win32.x86_64"/&gt;
&lt;plugin id="org.eclipse.core.resources.win32.x86_64"/&gt;

Added the following plug-ins:
&lt;plugin id="org.eclipse.e4.ui.swt.win32"/&gt;

Modified
&lt;plugin id="org.eclipse.equinox.security.win32.x86_64"/&gt; to 
&lt;plugin id="org.eclipse.equinox.security.win32"/&gt;

----
And I can now build a Win32 target on Linux!!!

sratz, Many thanks!!  That got me going in the right direction.

-Jeff

**Reply by laeubi** [reply]
As said before, the best is to use **sdk feature** in your product and **only add your plugin** to the product, you don't need to use "Add Required plug-ins"  PDE and Tycho will simply add what is required anyways!

**Reply by merks** [reply]
I'm not sure if it's possible to build a multi-platform product without using features because fragments need to be included via filters:

![image](https://github.com/user-attachments/assets/f4c9ebbb-e614-40fb-9fb1-781dd9d83faa)

If you don't reuse the platform features you will be managing this every time the platform changes something in this regard.

**Reply by laeubi** [reply]
I added support for that a while back so Tycho ignores non matching items, therefore this even works for plugin based products but it is cumbersome as you need to add all fragments yourself.

Beside that configuring **one** target (e.g. win32 as done here) does work to build on Linux. This also works in PDE as you can specific the ws/os/... in the target file.

**Reply by jclausius-sg** [reply]
@merks  - It does work. If you look at other replies (or view the discussion thru a browser), @sratz put me on the right track and I was able to solve the problem.

@laeubi - The app.product file is configured to be be based on plug-ins, and has been for more than 15 years.  Are you saying to change that option to be based on "plug-ins and features"?  Also, what SDK feature are you referring to?  I don't see something like 'org.eclipse.ide.sdk' or anything like that in the list of items I can add to app.product.

Also note, the project for this app has NO defined .target definition file.

-Jeff

**Comment by laeubi** [comment]
> Again, my guess is Tycho is looking at the build environment to determine what to download, and not what's specified in the "main" pom.xml.

Tycho can't know of anything not specified somewhere so it depends on your configuration (that we don't know), e.g. local build artifacts can have an influence, global maven settings and so on...
